src/client/dfuse/dfuse.h

```diff
@@ -138,6 +138,75 @@ struct dfuse_obj_hdl {
 	bool                      doh_kreaddir_finished;
 };
 
+/* Readdir support.
+ *
+ * Readdir is by far the most complicated component of dfuse as a result of the kernel interfaces,
+ * the dfs interface and the lack of kernel caching for concurrent operations.
+ *
+ * The kernel interface to readdir is to make a callback into dfuse to request a number of dentries
+ * which are then populated in a buffer which is returned to the kernel, each entry in the buffer
+ * contains the name of the entry, the type and some other metadata.  The kernel requests a buffer
+ * size and the dfuse can reply with less if it choses - 0 is taken as end-of-directory.  The
+ * length of the filenames affects the number of entries that will fit in the buffer.
+ *
+ * The dfs interface to reading entries is to call dfs_iterate() which then calls a dfuse callback
+ * with the name of each entry, dfuse then has to perform a lookup to get any metadata for the
+ * entry which it does after the iterate has completed.  DFS takes in a buffer size and max count
+ * which can be much larger than the 4k buffer the kernel uses.
+ *
+ * The kernel uses "auto readdir plus" to switch between two types of readdir, the plus calls
+ * return full stat information for each file including size (which is expensive to read) and
+ * it takes a reference for each entry returned so requires a hash table reference for each entry.
+ * The non-plus call just takes the name and mode for each entry so can do a lighter weight
+ * dfs lookup and does not need to do any per dentry hash table operations.
+ * For any directory the first call will be a plus type, subsequent entries will depend on if the
+ * application is doing stat calls on the dentries, "ls -l" will result in readdir plus being used
+ * throughout, "/bin/ls" will result in only the first call being plus.
```

Comment:
Using "ls -l" and then "/bin/ls" almost implies that using the full pathname results in different behavior

Reply:
it does.  "ls" is almost always aliased to "ls --color" or similar which actually does a stat to fetch the mode.  A common recommendation on HPC systems it to change the default shell setup to unalias ls for preciously this reason.

Reply:
Should the doc be explicit then? There are two differences, and reading the current wording, I (not an expert) don't know why the readdir behavior is different
- `ls -l` 
  - non-absolute path
  - argument `-l`
- `/bin/ls`
  - absolute path
  - no argument